source/common/upstream/upstream_impl.cc

```diff
@@ -919,9 +919,19 @@ ClusterImplBase::ClusterImplBase(
 
   auto socket_matcher = std::make_unique<TransportSocketMatcherImpl>(
       cluster.transport_socket_matches(), factory_context, socket_factory, *stats_scope);
-  info_ = std::make_unique<ClusterInfoImpl>(cluster, factory_context.clusterManager().bindConfig(),
-                                            runtime, std::move(socket_matcher),
-                                            std::move(stats_scope), added_via_api, factory_context);
+  auto& dispatcher = factory_context.dispatcher();
+  info_ = std::shared_ptr<const ClusterInfoImpl>(
+      new ClusterInfoImpl(cluster, factory_context.clusterManager().bindConfig(), runtime,
+                          std::move(socket_matcher), std::move(stats_scope), added_via_api,
+                          factory_context),
+      [&dispatcher](const ClusterInfoImpl* self) {
+        ENVOY_LOG(debug, "Schedule destroy cluster info {}", self->name());
+        dispatcher.movePost([raii_cluster = std::shared_ptr<const ClusterInfoImpl>(self)]() {
+          ENVOY_LOG(debug, "Destroying cluster info {}. This thread should be master thread.",
+                    raii_cluster->name());
```

Comment:
You may want to do an explicit:
```
  raii_cluster.reset();
```

Would it be possible to have some tests that verify that repo the original crash and thus verify that this change fixes the issue of objects being deleted from the wrong thread?

Reply:
added the explicit reset.

Reply:
Test case:https://github.com/lambdai/envoy-dai/commit/ea86eb1f2146b27048fc73ea0eccba2db36e90a6#diff-871b9f8bd15bf182038c744edcfb5895b9cc22895469f5dc4d8b2a8947602496R793

Reply:
This test reproduces the crash after the isThreadSafe() is added in Envoy::Grpc::AsyncStreamImpl::sendMessageRaw(). Thank you for the suggestion! Without thse assert(s), it is hard to prove it fixed.